File: scripts/scripts/utility/spotify_add.py
#!/usr/bin/python3
import base64
import json
import os
import sys

import requests

song_api_point = "https://api.spotify.com/v1/me/player"
save_api_point = "https://api.spotify.com/v1/me/tracks"
token_api_point = "https://accounts.spotify.com/api/token"


class Spotify:

    def loadDataFromFile(self):
        path = os.path.join(sys.path[0] + '/spotifyinfo.json')
        with open(path) as json_file:
            info_json = json.load(json_file)
            self.c_id = info_json['c_id']
            self.c_secret = info_json['c_secret']
            self.r_token = info_json['r_token']

    def getClientString(self):
        c_64 = self.c_id + ":" + self.c_secret
        c_64 = base64.b64encode(c_64.encode())
        c_64 = str(c_64)
        c_64 = c_64[2:-1]
        self.client_string = c_64

    def getToken(self):
        headers = {'Authorization': 'Basic ' + self.client_string}
        payload = {'grant_type': 'refresh_token',
                   'refresh_token': self.r_token}
        p = requests.post(token_api_point, headers=headers, data=payload)
        token_json = json.loads(p.text)
        self.my_token = str(token_json['access_token'])

    def getCurrentSong(self):
        params = {'market': 'ZA'}
        headers = {'Authorization': 'Bearer ' + self.my_token}
        self.song = requests.get(song_api_point, params=params,
                                 headers=headers).json()

    def addCurrentSongToLibrary(self):
        self.getCurrentSong()
        song = self.song['item']['uri']
        song_id = song[14:]
        params = {'ids': song_id}
        headers = {'Authorization': 'Bearer ' + self.my_token}
        requests.put(save_api_point, params=params, headers=headers)

    def __init__(self):
        self.loadDataFromFile()
        self.getClientString()
        self.getToken()


def main():
    spotify = Spotify()
    spotify.addCurrentSongToLibrary()


if __name__ == "__main__":
    main()

File: scripts/scripts/go/spotify_add.go
package main

import (
	"encoding/base64"
	"encoding/json"
	"io/ioutil"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"time"
)

var c = &http.Client{
	Timeout: 5 * time.Second,
}

type song struct {
	Item item `json:"item"`
}

type item struct {
	Title string `json:"name"`
	URI   string `json:"uri"`
}

type authToken struct {
	AccessToken string `json:"access_token"`
}

type client struct {
	ClientID     string `json:"c_id"`
	ClientSecret string `json:"c_secret"`
	RefreshToken string `json:"r_token"`
}

type spotify struct {
	accessToken   authToken
	client        client
	encodedClient string
	currentSong   song
}

func (s *spotify) GetClient() error {
	dir, err := filepath.Abs(filepath.Dir(os.Args[0]))
	if err != nil {
		return err
	}
	file := dir + "/spotifyinfo.json"
	jsonFile, err := os.Open(file)
	if err != nil {
		return err
	}
	byteValue, _ := ioutil.ReadAll(jsonFile)
	json.Unmarshal(byteValue, &s.client)
	// fmt.Println(s.client)
	return nil
}

func (s *spotify) EncodeClient() {
	myClient := s.client
	// fmt.Println(myClient)
	encode := myClient.ClientID + ":" + myClient.ClientSecret
	s.encodedClient = base64.StdEncoding.EncodeToString([]byte(encode))
	// fmt.Println(s.encodedClient)
}

func (s *spotify) GetNewToken() error {
	tokenAPI := "https://accounts.spotify.com/api/token"
	auth := url.Values{}
	auth.Set("grant_type", "refresh_token")
	auth.Set("refresh_token", s.client.RefreshToken)
	url := tokenAPI + "?" + auth.Encode()
	//fmt.Println(url)
	request, err := http.NewRequest(
		"POST",
		url,
		nil,
	)
	if err != nil {
		return err
	}
	authH := "Basic " + s.encodedClient
	request.Header.Set("Authorization", authH)
	request.Header.Set("content-type", "application/x-www-form-urlencoded")
	response, err := c.Do(request)
	if err != nil {
		return err
	}
	defer response.Body.Close()

	body, err := ioutil.ReadAll(response.Body)
	if err != nil {
		return err
	}
	var token authToken
	if err := json.Unmarshal(body, &token); err != nil {
		return err
	}
	s.accessToken = token
	return nil
}

func (s *spotify) GetCurrentSong() error {
	songAPI := "https://api.spotify.com/v1/me/player"
	request, err := http.NewRequest(
		"GET",
		songAPI,
		nil,
	)
	if err != nil {
		return err
	}
	request.Header.Set("Authorization", "Bearer "+s.accessToken.AccessToken)
	request.Header.Set("content-type", "application/x-www-form-urlencoded")
	response, err := c.Do(request)
	if err != nil {
		return err
	}
	defer response.Body.Close()
	body, err := ioutil.ReadAll(response.Body)
	if err != nil {
		return err
	}
	//fmt.Println(string(body))
	var currentSong song
	if err := json.Unmarshal(body, &currentSong); err != nil {
		return err
	}
	s.currentSong = currentSong
	//fmt.Println(currentSong)
	return nil
}

func (s *spotify) AddCurrentSongToLibrary() error {
	saveAPI := "https://api.spotify.com/v1/me/tracks"
	s.GetCurrentSong()
	currentSongID := s.currentSong.Item.URI[14:]
	auth := url.Values{}
	auth.Set("ids", currentSongID)
	url := saveAPI + "?" + auth.Encode()
	request, err := http.NewRequest(
		"PUT",
		url,
		nil,
	)
	if err != nil {
		return err
	}
	request.Header.Set("Authorization", "Bearer "+s.accessToken.AccessToken)
	request.Header.Set("content-type", "application/json")

	response, err := c.Do(request)
	if err != nil {
		return err
	}
	defer response.Body.Close()
	//fmt.Println(string(body))

	return nil
}

func main() {
	spotifyClient := spotify{}
	err := spotifyClient.GetClient()
	if err != nil {
		panic(err)
	}
	// fmt.Println(spotifyClient)
	spotifyClient.EncodeClient()
	err = spotifyClient.GetNewToken()
	if err != nil {
		panic(err)
	}

	err = spotifyClient.AddCurrentSongToLibrary()
	if err != nil {
		panic(err)
	}

}
